[DataLoader] Introduce MapMapDataPipe functional datapipe

MarcioPorto opened the pull request:
Summary:
As part of https://github.com/pytorch/pytorch/issues/57031, this PR adds the `MapMapDataPipe` functional datapipe for the `MapDataPipe` class.

Usage:
```
def fn(x):
    return x * 10

dp = CountingDataset(n=10)
dp.map(fn)
```

Differential Revision: D28394510

ejguan commented on torch/utils/data/dataset.py:
```diff
@@ -40,6 +41,27 @@ def __add__(self, other: 'Dataset[T_co]') -> 'ConcatDataset[T_co]':
     # See NOTE [ Lack of Default `__len__` in Python Abstract Base Classes ]
     # in pytorch/torch/utils/data/sampler.py
 
+    def __getattr__(self, attribute_name):
+        if attribute_name in Dataset.functions:
+            function = functools.partial(Dataset.functions[attribute_name], self)
+            return function
+        else:
+            raise AttributeError
+
+    @classmethod
+    def register_function(cls, function_name, function):
+        Dataset.functions[function_name] = function
+
+    @classmethod
+    def register_datapipe_as_function(cls, function_name, cls_to_register):
+        if function_name in Dataset.functions:
+            raise Exception("Unable to add DataPipe function name {} as it is already taken".format(function_name))
+
+        def class_function(cls, source_dp, *args, **kwargs):
+            return cls(source_dp, *args, **kwargs)
+        function = functools.partial(class_function, cls_to_register)
+        Dataset.functions[function_name] = function
```
We can get rid of the duplicate code for `IterableDataset` then.

VitalyFedyunin replied:
We can not, this `Dataset. functions` have to be different from `IterableDataset.functions` unless you want carefull extract super class here and use it.

ejguan replied:
For two `classmethod` functions, we can change `Dataset` and `IterableDataset`to make it happened, since we are using decorator of `functional_datapipe` to dispatch. Am I right?

MarcioPorto replied:
Yeah, that's correct. I can at least remove the two class methods using `cls`.

ejguan reviewed (approved):
Thanks for adding the first `MapDataPipe`. LGTM. One comment below.

MarcioPorto committed:
[DataLoader] Introduce MapMapDataPipe functional datapipe (#58258)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/58258

As part of https://github.com/pytorch/pytorch/issues/57031, this PR adds the `MapMapDataPipe` functional datapipe for the `MapDataPipe` class.

Usage:
```
def fn(x):
    return x * 10

dp = CountingDataset(n=10)
dp.map(fn)
```

Reviewed By: ejguan

Differential Revision: D28394510

fbshipit-source-id: c97396c59c4b3acdb12c3b4b05c8cff7dd5fcc50
